### ISSUE TITLE
A way to collect results upon exiting task group 

### ISSUE THREAD

**Issue by webknjaz** [agronholm/anyio#100]
In `asyncio`, there's `await asyncio.gather(..a..lot..of..tasks)`.

Is it possible to somehow emulate this with `anyio`?

I'm thinking of something like this maybe:
```python
async with create_task_group() as tg:
    await tg.spawn(t1)
    await tg.spawn(t2)

print(tg.results)  # t1_res, t2_res
```

**Comment by agronholm** [comment]
As with trio, and threads, you're supposed to use either a queue/memory channel, `deque`, `dict` or something else to store the results of the tasks. I have no plans of implementing any other means of task result retrieval.

**Comment by webknjaz** [comment]
@agronholm do you have any examples of how to do that? Also, are there any ways to `.spawn()` a task and get its instance returned?

**Comment by webknjaz** [comment]
Meanwhile, I've implemented an `asyncio.gather()` interface as follows: https://github.com/sanitizers/octomachinery/commit/839d4bfab74e5ebf5712a08383d77ef5185db301.

(Posting it here in case somebody else googles this issue)

**Comment by agronholm** [comment]
Just pass a dict, deque or whatever to the tasks and have the tasks put their results there?
What would you do with the task object? The only thing you could conceivably do is cancel it, but that is handled using cancel scopes in anyio (and trio).

**Comment by webknjaz** [comment]
Mostly I just want to collect results w/o having to add wrappers every time. This is why `asyncio.gather()` is so good: the boilerplate is hidden and I just use a simple call.

**Comment by smurfix** [comment]
The concept of structured concurrency doesn't have a "gather" for the same reason it doesn't have  the equivalent of "go" – the structured equivalent is a taskgroup, which waits for all its child tasks to end.

A ResultCollectingTaskgroup class is easy enough to program. You need an async context manager that holds the "real" taskgroup and a result list, a "spawn" method that allocates a slot in the result list and starts a helper, and a helper that catches the real result's return value (or error) in an Outcome and stores it in that slot.

I can do it easily if you ask nicely ;-) but no free time right now.

**Comment by webknjaz** [comment]
> I can do it easily if you ask nicely ;-)

@smurfix that would be great!

I think I'll stick to my wrappers for now: have you looked at them? Is there anything obviously wrong there?




> You need an async context manager that holds the "real" taskgroup and a result list

Why not inherit the task group class then and add a `gather` method there?

**Comment by webknjaz** [comment]
> What would you do with the task object?

Well, maybe if the task group was to expose a dict with all its subtask results, it'd be possible to use it for the result lookup...

**Comment by smurfix** [comment]
> Why not inherit the task group class

Because there is no such thing, that's a different class depending on your back-end.

**Comment by smurfix** [comment]
> have you looked at them? Is there anything obviously wrong there?

Well … you're doing way too much work. One taskgroup and N events is way cheaper than N taskgroups and N queues.

**Comment by smurfix** [comment]
https://gist.github.com/smurfix/d528056259496d1b7ada1c20b9aee6fb

**Comment by smurfix** [comment]
NB, that gist doesn't have events for waiting on single results. Easy enough to add if you want them, just another list with `anyio.Event` objects.

It also dies on first exception. While it is easy enough to teach `_run` to capture the function's result in an `Outcome` object, that'd swallow other exceptions; better to teach your code to replace exceptions which you consider benign with some reasonable return value.

**Comment by webknjaz** [comment]
> N taskgroups

Not sure what you mean here: in my case, one gather call needs one task group.

> gist.github.com/smurfix/d528056259496d1b7ada1c20b9aee6fb

Thanks for the gist! atm I still prefer `asyncio.gather`-like interface for readability reasons. Maybe your case would be reasonable for other use-cases than mine: I don't create a task group when calling my gather helpers so I just need one inside.

**Comment by webknjaz** [comment]
So now that the queues are gone (https://github.com/agronholm/anyio/issues/124 / https://github.com/agronholm/anyio/commit/5e8b5c551922bfec46db6a2d1a20121ab155f805) I need to figure out how to replace them with memory channels...

**Comment by smurfix** [comment]
```
try:
    from anyio import create_queue
except ImportError:
    from anyio import create_memory_object_stream as _cmos

    class Queue:
        def __init__(self, length=0):
            self._s, self._r = _cmos(length)

        def put(self, x):
            return self._s.send(x)

        def get(self):
            return self._r.receive()

        def qsize(self):
            return len(self._s._state.buffer)  # ugh

        def empty(self):
            return not len(self._s._state.buffer)  # ugh

        def __aiter__(self):
            return self

        def __anext__(self):
            return self._r.__anext__()

    def create_queue(length=0):
        return Queue(length)
```

**Comment by webknjaz** [comment]
Thanks for the shim!

**Comment by Tinche** [comment]
It's well within your rights to wontfix this, but I will chip in my $0.02 and say when I found out anyio/trio don't have good `gather` ergonomics my interest waned pretty significantly. In other words, I find `gather` in asyncio very, very good. Which is a little unfortunate because anyio/trio cancellation scopes are so awesome.

**Comment by smurfix** [comment]
This discussion already links to a reasonably simple implementation of `gather` on top of anyio.

**Comment by agronholm** [comment]
I'm not opposed to something like `.map()` but I'd like to coordinate with trio on this.

**Comment by phxnsharp** [comment]
Good discussion on this in the Trio repository, including [a much simpler implementation of gather](https://github.com/python-trio/trio/issues/2188#issuecomment-986161812) that would work just as well here.

**Comment by agronholm** [comment]
Yup, I'm favorable to a `gather()` implementation that would optionally collect the exceptions (`return_exceptions=True`).

**Comment by phxnsharp** [comment]
Now that I actually tried the code, there is an additional consideration. AnyIO and Trio both use function instances and `(*args, **kwargs)` when using functions like `nursery.start_soon()`. Passing multiple `(*args, **kwargs)` to a single `gather` function is difficult, and you cannot write an async lambda.

**Comment by agronholm** [comment]
What's wrong with `lambda: my_async_func(arg1, arg2)`?

**Comment by phxnsharp** [comment]
Turns out nothing aside from being limited to single function calls. Sorry, fairly new to Python :)

**Comment by graingert** [comment]
`lambda: my_async_func(arg1, arg2)` is technically forbidden in trio - as only a nursery is permitted to call an async function without an await

**Comment by agronholm** [comment]
So in the end, it doesn't matter if it's technically forbidden or not.
